rfcs/56-better-reflect.md

```diff
@@ -0,0 +1,261 @@
+they are supposed to stand for the requested type or not.
+
+For example
+
+```rust
+let a = MysteryType::new(); // suppose MysteryType implements Reflect
+let reflect_a: Box<dyn Reflect> = Box::new(a.clone()) as Box<dyn Reflect>;
+let dynamic_a: Box<dyn Reflect> = a.clone_value();
+// This more or less always panic
+assert_eq!((&*reflect_a).type_id(), (&*dynamic_a).type_id())
+```
+
+This is because multiple different things can pretend to be the same thing. In
+itself this shouldn't be an issue, but the current `Reflect` API is leaky, and
+you need to be aware of both this limitation and the underlying type to be able
+to use the `Reflect` API correctly.
+
+
+### `reflect_hash` and co. with `Dynamic***`
+
+The problem is not limited to `type_id`, but also extends to the `reflect_*`
+methods.
+
+```rust
+let a = MysteryType::new(); // suppose MysteryType implements Reflect
+let reflect_a: Box<dyn Reflect> = Box::new(a.clone()) as Box<dyn Reflect>;
+let dynamic_a: Box<dyn Reflect> = a.clone_value();
+let reflect_hash = reflect_a.reflect_hash();
+let dynamic_hash = dynamic_a.reflect_hash();
+// This more or less always panic if `MysteryType` implements `Hash`
+assert_eq!(reflect_hash, dynamic_hash)
+```
+
+
+## Proposition
+
+The problem stems from the duplicity of `dyn Reflect`. We want to rely on the
+underlying methods of the same thing to be unique.
+
+To solve this, we do not allow more than a single underlying type to "pass for"
+a concrete type in `Reflect`. This implies that the various `Dynamic***` types
+shouldn't implement `Reflect`.
+
+## User-facing explanation
+
+* `Reflect` maps 1-to-1 with an underlying type meaning that if a
+  `Box<dyn Reflect>` cannot be downcasted to `T`, you cannot build a `T` from
+  it.
+* There is a `ReflectProxy` enum you can use to construct any `Reflect` type
+  based on some dynamic structure. This can be used, for example, for
+  deserialization of types that implement `Reflect` but not `Deserialize`.
+  * You can "clone reflectively" a `Box<dyn Reflect>` by using
+    `Reflect::proxy_clone`. It returns a `ReflectProxy`.
+  * You can use `ReflectProxy::sidecast` to create a `Box<dyn Reflect>` from
+    a `ReflectProxy`. The created trait object's underlying type is a concrete
+    type you can downcast to.
+* `FromReflect` doesn't exist anymore. Instead, `Reflect::downcast` provides
+  now the same guarentees that `FromReflect::from_reflect` had. If you need to
+  get a `T` from a `ReflectProxy`, use a combination of
+  `ReflectProxy::sidecast_type<T>` and `downcast`.
+
+
+## Implementation strategy
+
+The `Reflect` API stays identical to the one we have today, at the exception of
+`clone_value`.
+
+We remove the `Dynamic***: Reflect` implementations and we introduce a
+`ReflectProxy` enum.
+
+```rust
+pub enum ReflectProxy {
+    Struct(DynamicStruct),
+    TupleStruct(DynamicTupleStruct),
+    Tuple(DynamicTuple),
+    List(DynamicList),
+    Map(DynamicMap),
+    AlreadyReflect(Box<dyn Reflect>),
```

Comment:
We could just keep this as `Value` (maybe `Primitive` or something similar). I think a user will be able to ascertain that a variant that returns `Box<dyn Reflect>` is already an implementor of `Reflect`. At least, that's my opinion.

Reply:
Well it couldn't be a struct unless it's marked as `#[reflect_value]`. So the only reflected types you should ever get in those cases are "value" types that don't fall into any of the other categories.

So I still think it should be consistent with that idea (and with the other enums). Again, this might just be a matter of preference, though.

Reply:
I think not. The user can build arbitrary struct, upcast it to `Box<dyn Reflect>` and insert it there. And I think it would be a reasonable thing to do (eg: fine grained control of when to convert which part of a `ReflectProxy` into `Reflect`).

Reply:
Okay I think I understand. So like in the code below?

```rust
let mut dyn_struct = DynamicStruct::default();
dyn_struct.insert("foo", Box::new(SomeStruct::new()));
let field_proxy = dyn_struct.field("foo");
assert!(matches!(field_proxy, ReflectProxy::AlreadyReflect(..)));
```

> Pretend `DynamicStruct` stores `Box<ReflectProxy>`

I think if that's the case, then I do agree that `AlreadyReflect` is a readable and reasonable naming choice. 

We could _maybe_ consider just `Reflect` or `Boxed` for the sake of keystrokes, but yeah `AlreadyReflect` is pretty good already.